src/System.Text.Primitives/System/Text/Encoders/Utf8.cs

```diff
@@ -48,296 +48,24 @@ public static OperationStatus FromUtf16(ReadOnlySpan<byte> source, Span<byte> de
-                        // Also, we need 7 chars reserve for the unrolled ansi decoding loop and for decoding of multibyte sequences
-                        byte* pStop = pSrc + availableBytes - 7;
-
-                        // Fast loop
-                        while (pSrc < pStop)
-                        {
-                            ch = *pSrc;
-                            pSrc++;
-
-                            if (ch > 0x7F)
-                                goto LongCode;
-
-                            bytesNeeded++;
-
-                            // 2-byte align
-                            if ((unchecked((int)pSrc) & 0x1) != 0)
-                            {
-                                ch = *pSrc;
-                                pSrc++;
-
-                                if (ch > 0x7F)
-                                    goto LongCode;
-
-                                bytesNeeded++;
-                            }
-
-                            // 4-byte align
-                            if ((unchecked((int)pSrc) & 0x2) != 0)
-                            {
-                                ch = *(ushort*)pSrc;
-                                if ((ch & 0x8080) != 0)
-                                    goto LongCodeWithMask16;
-                                pSrc += 2;
-                                bytesNeeded += 2;
-                            }
-
-                            // Run 8 characters at a time!
-                            while (pSrc < pStop)
-                            {
-                                ch = *(int*)pSrc;
-                                int chb = *(int*)(pSrc + 4);
-                                if (((ch | chb) & unchecked((int)0x80808080)) != 0)
-                                    goto LongCodeWithMask32;
-                                pSrc += 8;
-                                bytesNeeded += 8;
-                            }
-
-                            break;
-
-#if BIGENDIAN
-                    LongCodeWithMask32:
-                        // be careful about the sign extension
-                        ch = (int)(((uint)ch) >> 16);
-                    LongCodeWithMask16:
-                        ch = (int)(((uint)ch) >> 8);
-#else // BIGENDIAN
-                            LongCodeWithMask32:
-                            LongCodeWithMask16:
-                            ch &= 0xFF;
-#endif // BIGENDIAN
-                            pSrc++;
-                            if (ch <= 0x7F)
-                            {
-                                bytesNeeded++;
-                                continue;
-                            }
+                int indexOfFirstInvalidSequence = Utf8Util.GetIndexOfFirstInvalidUtf8Sequence(source, out int scalarCount, out int surrogatePairCount);
+                bytesNeeded = checked((scalarCount + surrogatePairCount) * 2 /* UTF-16 code units are 2 bytes in length */);
```

Comment:
nit: move the comment outside the code (address outside this PR). same with ToUTf32Length